[primTorch] Make `prims.collapse` a real prim

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #92017
* __->__ #91748

`prims.collapse` is currently just a plain python function wrapping
`prims.reshape`. This turns it into a real prim, and also factors out some of
the code duplicated with `_collapse_view_aten`.

cc @ezyang @mruberry @ngimel @Lezcano

peterbell10 committed:
[primTorch] Make `prims.collapse` a real prim

`prims.collapse` is currently just a plain python function wrapping
`prims.reshape`. This turns it into a real prim, and also factors out some of
the code duplicated with `_collapse_view_aten`.

[ghstack-poisoned]

peterbell10 committed:
Update on "[primTorch] Make `prims.collapse` a real prim"


`prims.collapse` is currently just a plain python function wrapping
`prims.reshape`. This turns it into a real prim, and also factors out some of
the code duplicated with `_collapse_view_aten`.

cc ezyang mruberry ngimel Lezcano

[ghstack-poisoned]

ezyang commented:
mmm, i can review this in a pinch but I think the other ccs will be better for it

peterbell10 committed:
Update on "[primTorch] Make `prims.collapse` a real prim"


`prims.collapse` is currently just a plain python function wrapping
`prims.reshape`. This turns it into a real prim, and also factors out some of
the code duplicated with `_collapse_view_aten`.

cc ezyang mruberry ngimel Lezcano

[ghstack-poisoned]

mruberry commented on torch/_prims/__init__.py:
```diff
@@ -1832,19 +1833,38 @@ def _as_strided_scatter_meta(
+    # Special-case for zero dimensional tensors
+    ndim = max(1, a.dim())
+    utils.validate_idx(ndim, start)
+    utils.validate_exclusive_idx(ndim, end)
```
Won't this prevent `collapse(t, 0, 0)` - is that invalid?

peterbell10 replied:
For whatever reason `aten.flatten(t, 0, 1)` flattens dims [0, 1] inclusive while `prims.collapse(t, 0, 1)` collapses dims `[0, 1)` exclusive. So the range [0, 0) is empty which is not allowed.

ngimel replied:
Why does behavior of `prims.collapse` and `aten.flatten` differ? We are under no obligation to follow `lax` names and behavior, or are we?

peterbell10 replied:
Looks like `prims.collapse` has been this way since the first prims PR #75095. cc @mruberry

mruberry replied:
We didn't intend to fix the prims' from day one, I was just modeling them on lax at the time to get the project started

mruberry commented on torch/_prims/__init__.py:
```diff
@@ -1832,19 +1833,38 @@ def _as_strided_scatter_meta(
 
-    new_shape = a.shape[0:start] + (dim_length,) + a.shape[end:]
-    return reshape(a, new_shape)
+    new_shape = _collapsed_shape(a.shape, start, end)
```
Is there a check that `end >= start`?

peterbell10 replied:
Added `_validate_collapse_args` which now handles this check.

mruberry commented on torch/_prims/__init__.py:
```diff
@@ -1832,19 +1833,38 @@ def _as_strided_scatter_meta(
+
+See collapse_view for the corresponding view operation.
+"""
+collapse = _make_prim(
```
Is there any testing for collapse? Should we add some tests to test_prims.py?

peterbell10 replied:
Test added

peterbell10 committed:
Update on "[primTorch] Make `prims.collapse` a real prim"


`prims.collapse` is currently just a plain python function wrapping
`prims.reshape`. This turns it into a real prim, and also factors out some of
the code duplicated with `_collapse_view_aten`.

cc ezyang mruberry ngimel Lezcano

[ghstack-poisoned]

peterbell10 commented:
@mruberry PTAL

peterbell10 commented:
@mruberry @ngimel @lezcano ping

lezcano commented on torch/_prims/__init__.py:
```diff
@@ -1335,25 +1355,12 @@ def _collapse_view_meta(a: TensorLikeType, start: int, end: int) -> TensorLikeTy
         raise ValueError(msg)
 
-    if new_strides is None:
-        return a.view(new_shape)
-    else:
-        return a.as_strided(new_shape, new_strides, a.storage_offset())
+    assert new_strides is not None
```
Let's just raise within `collapse_view_helper` and never return `None`.

peterbell10 replied:
`_collapse_view_helper` is used by `_refs.flatten` which returns a copy when `_collapse_view_helper` returns `None`.

lezcano replied:
oof. ok

peterbell10 committed:
Update on "[primTorch] Make `prims.collapse` a real prim"


`prims.collapse` is currently just a plain python function wrapping
`prims.reshape`. This turns it into a real prim, and also factors out some of
the code duplicated with `_collapse_view_aten`.

cc ezyang mruberry ngimel Lezcano

[ghstack-poisoned]
